src/Compiler.Dynamic/SystemWebCompilation.cs

```diff
@@ -122,11 +124,52 @@ private IEnumerable<BaseTemplateParser> GetParsersToCompile(IEnumerable<string>
             if (ReferenceEquals(stack.Peek(), currentPath))
             {
                 stack.Pop();
+                parser.Parse();
                 yield return parser;
             }
         }
     }
 
+    private IEnumerable<string> GetDependencyPaths(TemplateParser parser)
+    {
+        if (!_dependencies.TryGetValue(parser.CurrentVirtualPathString, out var dependentPaths))
+        {
+            var controlDependencyParser = GetConcreteDependencyParser(parser);
+            controlDependencyParser.Init(parser.CurrentVirtualPath);
+            controlDependencyParser.WebFormsFileProvider = parser.WebFormsFileProvider;
+            var result = controlDependencyParser.GetVirtualPathDependencies();
+            _dependencies[parser.CurrentVirtualPathString] = dependentPaths = result;
+        }
+
+        if (dependentPaths is not null)
+        {
+            foreach (string virtualPathString in dependentPaths)
+            {
+                yield return virtualPathString;
+            }
+        }
+    }
+
+    private static DependencyParser GetConcreteDependencyParser(TemplateParser parser)
```

Comment:
Please integrate this into the PageCompilationOptions and determine if we should start with the dependency parser and expose a way from a dependency parser to get the full parser. I don't want to see so many type checks going on.

Reply:
implemented that way.. please take a look